src/Swarm/Game/State.hs

```diff
@@ -850,6 +864,61 @@ wakeUpRobotsDoneSleeping = do
       let aliveRids = filter (`IM.member` robots) rids
       internalActiveRobots %= IS.union (IS.fromList aliveRids)
 
+      -- These robots' wake times may have been moved "forward"
+      -- by "wakeWatchingRobots".
+      clearWatchingRobots aliveRids rids
+
+-- | Clear the "watch" state of all of the
+-- awakened robots
+clearWatchingRobots ::
+  (Has (State GameState) sig m, Foldable t) =>
+  t Int ->
+  [RID] ->
+  m ()
+clearWatchingRobots aliveRids rids = do
+  robotsWatching %= (`S.difference` S.fromList rids)
+  forM_ aliveRids $ \rid -> robotMap . at rid . _Just . watchedLocations .= mempty
+
+-- | Iterates through all of the currently "wait"-ing robots,
+-- and moves forward the wake time of the ones that are watching this location.
+--
+-- NOTE: Clearing the "wake time" ticks entries in "internalWaitingRobots"
+-- is handled by "wakeUpRobotsDoneSleeping" in State.hs
+wakeWatchingRobots :: Has (State GameState) sig m => Location -> m ()
+wakeWatchingRobots loc = do
+  currentTick <- use ticks
+  waitingMap <- use waitingRobots
+  rMap <- use robotMap
+
+  let watchingBots =
+        S.filter (botHasWatchedLoc rMap)
+          . S.unions
+          . map S.fromList
+          $ M.elems waitingMap
+
+      newWakeTime = currentTick + 1
+      newInsertions = M.singleton newWakeTime $ S.toList watchingBots
+
+      filteredWaitingMap =
+        M.filter (not . null) $
+          M.map (filter (`S.notMember` watchingBots)) waitingMap
+
+  -- NOTE: There are two "sources of truth" for the waiting state of robots:
+  -- 1. In the GameState via "internalWaitingRobots"
+  -- 2. In each robot, via the CESK machine state
```

Comment:
This was a gotcha for me, after having plumbed just the waking logic for `internalWaitingRobots` but observing that the robots still did not wake up.

Would it be feasible to have a single source of truth?  The irony of now introducing both a `robotsWatching` in `GameState` and `watchedLocations` in each `Robot` is not lost on me.

Reply:
I've at least added some comments so I guess I can live with the the dual sources of truth for now.